### PR TITLE
fix(runtime): recover pre-3.0 live session paths

### PR DESCRIPTION
**Pull request by JeongJaeSoon**
## Summary

- reproduce a real pre-3.0 Claude shell snapshot and loaded absolute hook command after their `1.9.0` cache directory is removed
- repair recoverable missing legacy semver paths as marked compatibility shims through a hidden link to the newest real binary
- harden runtime, manifest, shell-init, and setup-shell resolvers against marked, marker-less symlink, and malformed-version rollback cycles
- document the host lifecycle boundary where an already-running legacy session cannot self-heal

## Reproduction

Before #143, Claude shell snapshots and loaded hook commands could retain an absolute path such as:

```text
~/.claude/plugins/cache/agent-guard/agent-guard/1.9.0/bin/agent-guard
```

After the host deletes `1.9.0`, both `command -v agent-guard` in that snapshot and an already-loaded hook command fail with status 127. The earlier clean-home model did not reproduce this because it copied the new resolver implementation into a synthetic old version and invoked that implementation before deletion.

## Root cause

The stable `current` link and current resolver are entered only after a new binary runs. A command already loaded by an older host points directly at the deleted executable and cannot reach that resolver. The host exposes no plugin lifecycle callback at the exact instant it removes the old cache directory.

The first compatibility implementation also pointed repaired shims back through `current`. After rollback, a pre-fix resolver could make that shim executable on its first invocation, select the higher shim version on its second invocation, and create `current -> shim -> current`.

## Change

- repair an exact missing `CLAUDE_PLUGIN_ROOT` / `PLUGIN_ROOT` semver path when a current manifest resolver is invoked
- on current `SessionStart`, `shell-init`, `setup-shell`, or `version`, parse only plain non-symlink Claude shell snapshots as text, never source them, and enforce file, byte, line, candidate, and version caps
- honor `CLAUDE_CONFIG_DIR/shell-snapshots` without requiring or mixing in `HOME`
- recreate only a missing direct-child older semver path beneath the same plugin cache
- create the compatibility marker as an atomic directory, while continuing to recognize earlier file markers
- symlink new compatibility executables through `.agent-guard-real-current`, which tracks only the newest real version and never points back through `current`
- migrate every exact old managed shape (real semantic-version directory, real `bin` and marker, final binary symlink, and `current/bin/agent-guard` target) during current-binary startup
- preserve a pre-existing real `current` directory only when its `bin` and executable are also real cache entries
- keep all runtime/manifest/shell-init/setup-shell fallbacks from treating compatibility symlinks as installed releases even if their marker is removed
- reject symlinked version roots, `bin` directories, final candidates, stale `current` targets, and unsafe exact host-root fallbacks across runtime, Codex/Claude manifests, shell-init, and setup-shell
- reject malformed numeric lookalikes such as trailing-dot version directories
- avoid a duplicate bounded snapshot scan when a generated shell-init snippet probes the resolved binary version
- preserve real directories during concurrent repair and document rollback, read-only, bounded-out, no-snapshot, and no-current-binary limits

## Latest-main refresh

- merged current `main` at `16b116a1817bf3b14156e986d7aef75799e1a021` into this branch with a signed merge commit
- revalidated after merged PRs #153, #161, and #162; none duplicates this legacy-session migration
- PR #162 moved marketplace SHA pinning to post-merge `make submission-artifact SHA=<merged-main-sha>`, so this Draft PR does not add a manual SHA pin commit

## Related work

- follow-up to #143
- #150, merged #153, and #156 remain adjacent work, not duplicate legacy-session migration

## Verification

- two independent sub-agent security/POSIX/false-positive reviews — found the migration cap, cache-parent symlink escapes, stale-current resolver paths, and unsafe real-`current` exception; the final refreshed reviews reported no remaining actionable findings
- `make test` — 830 passed, 0 failed
- `make smoke-test` — passed
- `scripts/validate-plugin-layout.sh` — passed
- `make submission-check` — passed
- `sh -n plugins/agent-guard/bin/agent-guard`
- `dash -n plugins/agent-guard/bin/agent-guard`
- `sh -n tests/run.sh`
- `dash -n tests/run.sh`
- `jq -e . plugins/agent-guard/hooks.json`
- `jq -e . plugins/agent-guard/hooks/hooks.json`
- `git diff --check` — passed

The Makefile has no `fmt` target.

## Recovery boundary / not covered

An old session cannot recover until some current binary executes. Starting a current session or invoking a real installed `agent-guard setup-shell` triggers migration. If the cache is read-only, symlinks are unavailable, no matching plain snapshot/host-root record remains, or the bounded snapshot scan excludes the record, the legacy session must be restarted.

After rollback removes the real target behind a migrated shim, the shim intentionally remains non-executable so a pre-fix resolver cannot select it. The dangling shim cannot bootstrap its own repair; invoke a real installed version or restart the host to run the fallback resolver. Cache rollback and migration must not run concurrently because POSIX `ln` cannot conditionally replace the managed symlink against a cache manager rewriting the same entry.

No live installed Claude Code or Codex host lifecycle E2E was executed; deterministic fixtures cover the real removed-directory snapshot and loaded absolute hook path.

### CODE REVIEW

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

@codex review

File: plugins/agent-guard/bin/agent-guard [hunk 1]
```diff
@@ -89,6 +107,138 @@ refresh_plugin_current_link || :
...
+create_legacy_plugin_shim() {
+  legacy_root=${1:-}
+  [ -n "$PLUGIN_CACHE_BASE" ] && [ -n "$legacy_root" ] || return 0
+  [ "${legacy_root%/*}" = "$PLUGIN_CACHE_BASE" ] || return 0
+  legacy_version=${legacy_root##*/}
+  is_semver_dir "$legacy_version" || return 0
+  # Compatibility shims are only for removed pre-current releases. A shim with
+  # a newer version could win an older manifest resolver and point back through
+  # `current`, creating a current -> shim -> current cycle.
+  semver_less_than "$legacy_version" "$PLUGIN_CACHE_VERSION" || return 0
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard:162]: Addressed in signed commit 6feb722. All four Claude and all four Codex manifest resolvers now exclude directories carrying the reserved compatibility-shim marker before version selection. The regression reproduces a dangling 3.0.1 current, a marked 2.9.0 shim, and a real 2.0.0 rollback, then runs each host resolver twice and requires current to remain on 2.0.0 with hook input preserved. make test passes 563/0.

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

@codex review

File: plugins/agent-guard/bin/agent-guard [hunk 2]
```diff
@@ -89,6 +107,138 @@ refresh_plugin_current_link || :
...
+  done
+}
+
+# Older Claude shell snapshots can be the only remaining record of a removed
+# cache version. Parse only plain regular files as text (never source them), and
+# bound each invocation by candidate files, total bytes, and versions.
+repair_legacy_shell_snapshots() {
+  [ -n "$PLUGIN_CACHE_BASE" ] && [ -n "${HOME:-}" ] || return 0
+  [ -x "$STABLE_BIN" ] || return 0
+  snapshot_dir=$HOME/.claude/shell-snapshots
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard]: Implemented in signed commit 42b294a. Snapshot discovery now prefers CLAUDE_CONFIG_DIR/shell-snapshots, works with HOME unset, and a regression test verifies that a configured directory does not also scan a decoy HOME/.claude tree. I also guarded the HOME-derived private gitleaks fallback so an empty directory cannot become /gitleaks.

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

@codex review

File: README.md [hunk 3]
```diff
@@ -403,8 +403,27 @@ agent-guard setup-shell --no-command-wrapping  # persistent opt-out
 For plugin installs, every execution refreshes a sibling
 `current/bin/agent-guard` symlink and `setup-shell` records only that stable
 path. Hooks and shell snippets use `current` first, then select the newest
-installed semantic-version directory if symlinks are unavailable. This keeps an
-already-running session valid when the host removes an older cache directory.
+installed semantic-version directory if symlinks are unavailable. A session
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply README.md]: Clarified in signed commit 5552a75. The README now states that semantic-version fallback applies only to new resolver invocations and cannot repair an already-loaded legacy absolute path; that path still requires the symlink-backed compatibility shim and a current binary invocation described immediately below.

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

@codex review

**Comment by JeongJaeSoon** [comment]
## Validation (2026-08-03)

- two independent sub-agent security/POSIX/false-positive reviews — actionable migration-cap and symlink-boundary findings were incorporated; both final refreshed reviews reported no remaining actionable findings
- `make test` — passed: 830, failed: 0
- `make smoke-test` — passed
- `scripts/validate-plugin-layout.sh` — passed
- `make submission-check` — passed
- `sh -n plugins/agent-guard/bin/agent-guard` — passed
- `dash -n plugins/agent-guard/bin/agent-guard` — passed
- `sh -n tests/run.sh` — passed
- `dash -n tests/run.sh` — passed
- both hook manifests parsed with `jq -e .`
- `git diff --check` — passed
- signed merge, hardening, and follow-up commits verified locally; GitHub verification checked separately after push

### GitHub evidence

- commits `628c520e72c832ba24d504086e22939620bc8016`, `f50212da20eed2cdf2847f27eef1e9f99f3d223f`, `6d4ecbce542ea9031fae1907aeb5dba12e524d01`, and `382dbfcf467c96e30112071ed408f1065262b92a` — GitHub `verified: true` (`reason: valid`)
- CI run `30804910161` actual logs — Ubuntu 825 passed / 0 failed; macOS 777 passed / 0 failed; Ubuntu real-gitleaks smoke-test completed successfully; lint passed
- HOL run `30804909896` actual logs — critical 0, high 0, medium 1, low 0, final score 96/100
- Plugin Validation run `30804909930` actual logs — Codex, Claude, marketplace, and release-archive layouts passed; submission readiness passed
- unresolved inline review threads — 0 after replying to and resolving the Codex P2 migration-cap finding

The regressions cover a real removed `1.9.0` snapshot and loaded hook path, an existing current-backed shim with a pre-fix resolver invoked twice after rollback, migration of all 129 cycle-capable shims, hidden real-version targeting, safe and unsafe real `current` layouts, marker and marker-less rollback cycles, root/bin/final symlink exclusion across runtime/manifests/shell-init/setup-shell, stale read-only `current`, trailing-dot lookalikes, bounded FIFO/symlink/oversized snapshot handling, concurrent repair, `CLAUDE_CONFIG_DIR`, and a single shell-init snapshot scan.

## Not covered

- no live installed Claude Code or Codex host lifecycle E2E
- an old process cannot self-heal until a current binary runs; read-only cache, missing snapshots/host roots, unavailable symlinks, and bounded-out records require a session restart
- after rollback removes the migrated shim's real target, the dangling shim cannot bootstrap its own repair; invoke a real installed version or restart the host
- cache rollback and compatibility migration must not run concurrently
- no post-merge marketplace artifact; PR #162 intentionally generates it from the merged main SHA

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

@codex review

File: plugins/agent-guard/bin/agent-guard [hunk 4]
```diff
@@ -89,6 +108,142 @@ refresh_plugin_current_link || :
...
+create_legacy_plugin_shim() {
+  legacy_root=${1:-}
+  [ -n "$PLUGIN_CACHE_BASE" ] && [ -n "$legacy_root" ] || return 0
+  [ "${legacy_root%/*}" = "$PLUGIN_CACHE_BASE" ] || return 0
+  legacy_version=${legacy_root##*/}
+  is_semver_dir "$legacy_version" || return 0
+  # Compatibility shims are only for removed pre-current releases. A shim with
+  # a newer version could win an older manifest resolver and point back through
+  # `current`, creating a current -> shim -> current cycle.
+  semver_less_than "$legacy_version" "$PLUGIN_CACHE_VERSION" || return 0
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard:162]: Fixed in 6d4ecbce542ea9031fae1907aeb5dba12e524d01. The original fixture did copy the current resolver into the rollback release, so it could not exercise this pre-fix cycle.

The follow-up now:

- routes newly created compatibility shims through a hidden link that tracks only the newest real version;
- migrates only the exact previously managed shape (marker present, semantic-version directory, final binary symlink, and the old `current/bin/agent-guard` target), with a 128-candidate startup cap;
- preserves a pre-existing real `current` directory instead of replacing or sweeping it; and
- seeds an existing old-style current-backed shim, removes the new release, and invokes a faithful pre-fix resolver twice. Both rounds keep `current` on the real 2.0.0 release and the compatibility shim remains non-executable, so no `current -> shim -> current` cycle forms.

The documented recovery boundary is explicit: after rollback removes the hidden link's real target, the dangling legacy shim cannot bootstrap itself; a real installed version or host restart must run the fallback resolver. Cache rollback must not run concurrently with migration.

Validation on the pushed commit:

- `make test`: 823 passed, 0 failed
- `make smoke-test`: passed
- `scripts/validate-plugin-layout.sh`: passed
- `make submission-check`: passed
- `sh -n` and `dash -n` for the runtime and test script: passed
- `git diff --check`: passed

An independent sub-agent reproduced the original cycle, reviewed the updated migration and boundary cases, and reported no remaining actionable finding.

**Comment by JeongJaeSoon** [comment]
@codex review

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

File: plugins/agent-guard/bin/agent-guard [hunk 5]
```diff
@@ -55,30 +55,79 @@ is_semver_dir() (
...
+  # the snapshot or host-root reference that originally created it is gone.
+  # Marker-less and non-symlink cache entries are never rewritten.
+  if [ "$(readlink "$compat_real" 2>/dev/null)" = "$latest" ]; then
+    old_stable="$PLUGIN_CACHE_BASE/current/bin/agent-guard"
+    compat_real_bin="$compat_real/bin/agent-guard"
+    migration_candidates=0
+    max_migration_candidates=128
+    for legacy_bin in "$PLUGIN_CACHE_BASE"/*/bin/agent-guard; do
+      migration_candidates=$((migration_candidates + 1))
+      [ "$migration_candidates" -le "$max_migration_candidates" ] || break
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard]: Fixed in signed commit 382dbfcf467c96e30112071ed408f1065262b92a.\n\nThe startup cap is removed: every exact old managed shape (real semantic-version root, real bin and marker directories, final binary symlink, and the old current/bin/agent-guard target) is migrated before rollback. The regression now creates 129 current-backed shims, requires all 129 to point through .agent-guard-real-current with zero old targets left, removes 3.0.1, and invokes a faithful pre-fix 2.0.0 resolver twice. Both invocations keep current on the real rollback release.\n\nThe same follow-up rejects symlinked root/bin/final candidates and stale current targets across runtime, both host manifests, shell-init, and setup-shell. Final validation: make test 830 passed / 0 failed; smoke, plugin layout, submission readiness, sh/dash syntax, and git diff --check all passed. Two independent refreshed reviews reported no remaining actionable findings.

**Comment by JeongJaeSoon** [comment]
@codex review

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

**Comment by JeongJaeSoon** [comment]
Closing this as not planned. The affected case is a pre-3.0 live session retaining an absolute path into a plugin-cache version after that directory has been removed. It is real, but exceptionally narrow, and recovery is available by starting a new session or invoking a current installed Agent Guard binary.

The proposed compatibility layer is disproportionate to that impact: it adds legacy snapshot parsing and shim lifecycle handling across runtime, manifests, shell-init, and setup-shell (+1,169/-71). The latest head also still has an unresolved P1 around regular-file shim markers that can recreate a `current -> shim -> current` rollback cycle and disable hooks.

Given the maintenance and security risk, I do not want to merge this compatibility surface for a rare, recoverable case. The branch is being kept for reference. If concrete user reports show that restart/current-binary recovery is insufficient, we can revisit the problem with a narrower design.
